### PR TITLE
MM-56877: Reduce usages of SessionHasPermissionToChannelByPost

### PR DESCRIPTION
**Pull request by agnivade**
1. For file endpoints, a recent optimization added the ChannelID column
to be part of the fileinfo table. Therefore, we can skip the postID
and directly use the channelID.
2. For post endpoints, we reorder the sequence of calls such that
we get the channelID first, and use it to check the permission of the channel
rather than query the long way around by joining with the posts table
in the permissions query.

   The benefit is that SessionHasPermissionToChannel is cache-backed.
So in the happy path, we save a DB call. Because GetSinglePost anyways
needed to be called.
And in the bad path, we replace it with a more efficient call. Because
SessionHasPermissionToChannel is cache-backed, so effectively we
are replacing SessionHasPermissionToChannelByPost with GetSinglePost.

While here, we also do some other optimizations:

3. Pre-populate the channelID while saving the reaction, so that
we don't need to query the posts table for every single reaction save.
4. Remove unnecessary goroutine spawning for publishing reaction events,
because anyways those are asynchronous.

https://mattermost.atlassian.net/browse/MM-56877

```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
@jupenur - (NOT URGENT) I wanted to run this by you first to see if there are any security impacts here since it has to do with permission checks. Let me know.

**Review by jupenur — COMMENTED** [review]
No security concerns, but you should coordinate with ABC on the channel bookmarks change: https://github.com/mattermost/mattermost/pull/25449

They are adding a new way to attach files, which means they touch the file endpoints as well, so there will be conflicts, and also the new file attachment type has its own rules on who can access it.

**Comment by agnivade** [comment]
Ok thanks! Let me run this through dev review first.

File: server/channels/api4/file.go [hunk 1]
```diff
@@ -460,7 +460,14 @@ func getFile(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 	audit.AddEventParameterAuditable(auditRec, "file", info)
 
-	if info.CreatorId != c.AppContext.Session().UserId && !c.App.SessionHasPermissionToChannelByPost(*c.AppContext.Session(), info.PostId, model.PermissionReadChannelContent) {
+	perm := false
+	if info.ChannelId != "" {
```
Comment (streamer45) [inline server/channels/api4/file.go:464]: Wondering if this could be simplified further. Are there are instances in which the channel ID would be empty at this point? I suppose only for non-attached files but then it wouldn't make sense to check the post either.

**Review by streamer45 — APPROVED** [review]
Nice improvements :+1:

**Comment by agnivade** [comment]
Ok, so I looked further into it and there are some problematic inconsistencies in the way how `SessionHasPermissionToChannel` and `SessionHasPermissionToChannelByPost` is handled.

If you pass empty postID to `SessionHasPermissionToChannelByPost`, but if the session is of a sysadmin, it returns true. If you pass an empty channelID to `SessionHasPermissionToChannel`, it always returns false because it checks for empty channelID first.

Initially, I thought there was a difference due to the file not being attached to a channel, but then when Claudio raised the point, I looked at found out that even postID was empty but still the check was returning true!

**Tldr**:

The general idea behind `SessionHasPermissionToChannelByPost` is to cascade checks one after the other, never to return false, until the very last check.

And the general idea behind `SessionHasPermissionToChannel` is to validate some things, and possibly return an early false without checking for other things.

This difference leads to one returning false for an invalid input, and the other cascading to a point where session.Unrestricted returns true.

With these changes, I was able to make the new auth check pass as well:

```go
func (a *App) SessionHasPermissionToChannel(c request.CTX, session model.Session, channelID string, permission *model.Permission) bool {
	// if channelID == "" {
	// 	return false
	// }

	ids, err := a.Srv().Store().Channel().GetAllChannelMembersForUser(session.UserId, true, true)
	var channelRoles []string
	if err == nil {
		if roles, ok := ids[channelID]; ok {
			channelRoles = strings.Fields(roles)
			if a.RolesGrantPermission(channelRoles, permission.Id) {
				return true
			}
		}
	}

	channel, appErr := a.GetChannel(c, channelID)
	// if appErr != nil && appErr.StatusCode == http.StatusNotFound {
	// 	return false
	// }

	// if session.IsUnrestricted() {
	// 	return true
	// }

	if appErr == nil && channel.TeamId != "" {
		return a.SessionHasPermissionToTeam(session, channel.TeamId, permission)
	}

	return a.SessionHasPermissionTo(session, permission) // session.IsUnrestricted is checked here
}
```

The question is - do I make these changes and see what breaks?

Our requirements are:
1. Return true for sysadmin always.
2. Return false for invalid input, and if not a sysadmin.
3. Cascade checks as well.

I am willing to spend some time and fix this because it has performance implications as well, and we have not been very consistent about our logic.

cc @jupenur @streamer45

**Comment by agnivade** [comment]
Ok, even the tests are inconsistent. There's a test which verifies that we get a 403 error for a sysadmin client and a bad channelID. Yet, we get no error for a bad postID and a sysadmin!

In any case, one test or the other should be changed.

**Review by isacikgoz — APPROVED** [review]

**Comment by agnivade** [comment]
Ok, so I changed the tests to return 403 if there's no channelID or postID.

**Comment by agnivade** [comment]
@streamer45 @isacikgoz - Could I please ask you to review this new commit: https://github.com/mattermost/mattermost/pull/26239/commits/84291acedf64c34c1b97448f5fd4855d204f6d45 that got added? Thanks.

**Comment by agnivade** [comment]
@enahum - Just a heads up that this PR optimizes the permissions check in the fileInfo APIs to check for channel rather than post. You are already doing a similar thing in https://github.com/mattermost/mattermost/pull/25449 so there might be a conflict, so just a heads up in that case.

**Comment by enahum** [comment]
Thanks for the heads up @agnivade if there are conflicts I'll make sure to address it

**Review by isacikgoz — APPROVED** [review]
Looks good still!

**Review by streamer45 — APPROVED** [review]
Nice work!

**Comment by agnivade** [comment]
E2E tests look good. Merging!
